ruby/lib/plus_codes/open_location_code.rb

```diff
@@ -39,40 +37,68 @@ def full?(code)
     #
     # @param latitude [Numeric] a latitude in degrees
     # @param longitude [Numeric] a longitude in degrees
-    # @param code_length [Integer] the number of characters in the code, this excludes the separator
+    # @param code_length [Integer] the number of characters in the code, this
+    # excludes the separator
     # @return [String] a plus+codes
     def encode(latitude, longitude, code_length = PAIR_CODE_LENGTH)
-      raise ArgumentError,
-      "Invalid Open Location Code(Plus+Codes) length: #{code_length}" if invalid_length?(code_length)
-
-      if code_length > MAX_CODE_LENGTH
-        code_length = MAX_CODE_LENGTH
+      if invalid_length?(code_length)
+        raise ArgumentError, 'Invalid Open Location Code(Plus+Codes) length'
       end
+
+      code_length = MAX_CODE_LENGTH if code_length > MAX_CODE_LENGTH
       latitude = clip_latitude(latitude)
       longitude = normalize_longitude(longitude)
       latitude -= precision_by_length(code_length) if latitude == 90
 
-      lat = (latitude + 90).to_r
-      lng = (longitude + 180).to_r
+      final_lat_precision = 8000 * 5**(MAX_CODE_LENGTH - PAIR_CODE_LENGTH)
+      final_lng_precision = 8000 * 4**(MAX_CODE_LENGTH - PAIR_CODE_LENGTH)
 
-      digit = 0
+      # This algorithm starts with the least significant digits, and works it's
+      # way to the front of the code.
       code = ''
-      while digit < code_length
-        lat, lng = narrow_region(digit, lat, lng)
-        digit, lat, lng = build_code(digit, code, lat, lng)
-        code << SEPARATOR if (digit == SEPARATOR_POSITION)
+      if code_length > PAIR_CODE_LENGTH
+        # Multiply the decimal part of each coordinate by the final precision so
+        # we can treat them as integers. Round them off to 6 decimal places
+        # before converting to integers to avoid floating point rounding errors.
+        lat_p =
+          (((latitude - latitude.floor) * final_lat_precision * 1e6).round /
```

Comment:
I'm guessing 6 places was chosen because that's the max before we run into 32-bit integer overflow issues? Most languages (all?) support 64-bit integers - should we consider upping to maybe 12 digits to increase the precision?

Reply:
Refactored, this goes away.